src/charm.py

```diff
@@ -561,29 +575,75 @@ def _on_mongod_pebble_ready(self, event) -> None:
 
         except (PathError, ProtocolError, MissingSecretError) as e:
             logger.error("Cannot initialize workload: %r", e)
-            event.defer()
-            return
+            return False
 
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("mongod", self._mongod_layer, combine=True)
-        container.add_layer("log_rotate", self._log_rotate_layer, combine=True)
+        return True
+
+    def __configure_layers(self, container: Container) -> bool:
+        """Configure the layers of the container."""
+        modified = False
+        current_layers = container.get_plan()
+        new_layers = {
+            Config.SERVICE_NAME: self._mongod_layer,
+            "log_rotate": self._log_rotate_layer,
+        }
         if self.is_role(Config.Role.CONFIG_SERVER):
-            container.add_layer("mongos", self._mongos_layer, combine=True)
+            new_layers["mongos"] = self._mongos_layer
 
+        # Add Pebble config layers missing or modified
+        for layer_name, definition in new_layers.items():
+            for service_name, service in definition.services.items():
+                if current_layers.services.get(service_name) != service:
+                    modified = True
+                    logger.debug(f"Adding layer {service_name}.")
+                    container.add_layer(layer_name, definition, combine=True)
```

Comment:
`replan` will restart all changed services.
I'm mainly trying to be lazy here and avoid any unnecessary changes, and it seems it was needed because without that, I would quickly encounter some start/restart issues.

Reply:
>  it seems it was needed because without that, I would quickly encounter some start/restart issues.

This is very surprising. 🤔 This means that it would restart/start even if the plan was the same?

Another things that I would expect `container.add_layer(layer_name, definition, combine=True)` to combine the layers automatically so that you don't have to check/worry about adding duplicates

Reply:
I think it might be the difference between `1m0s` and `1m` that was doing it but I definitely ended up having some issues here otherwise I wouldn't have gone that way.

Reply:
There:
```
unit-mongodb-k8s-1: 12:02:48 ERROR unit.mongodb-k8s/1.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/./src/charm.py", line 1789, in <module>
    main(MongoDBCharm)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/main.py", line 551, in main
    manager.run()
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/main.py", line 530, in run
    self._emit()
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/main.py", line 519, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/main.py", line 147, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/framework.py", line 348, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/framework.py", line 860, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/framework.py", line 950, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/./src/charm.py", line 747, in _on_start
    if not self.__can_charm_start():
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/./src/charm.py", line 711, in __can_charm_start
    self._configure_container(container)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/./src/charm.py", line 625, in _configure_container
    self.__configure_layers(container)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/./src/charm.py", line 605, in __configure_layers
    container.replan()
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/model.py", line 2259, in replan
    self._pebble.replan_services()
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/pebble.py", line 2129, in replan_services
    return self._services_action('replan', [], timeout, delay)
  File "/var/lib/juju/agents/unit-mongodb-k8s-1/charm/venv/ops/pebble.py", line 2226, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "pbm-agent" (cannot start service: exited quickly with code 1)
----- Logs from task 1 -----
2024-10-09T10:02:48Z INFO Service "mongod" already started.
----- Logs from task 2 -----
2024-10-09T10:02:48Z INFO Service "mongodb_exporter" already started.
----- Logs from task 3 -----
2024-10-09T10:02:48Z INFO Most recent service output:
    2024/10/09 10:02:48 Exit: connect to the node: ping: connection() error occurred during connection handshake: auth error: sasl conversation error: unable to authenticate using mechanism "SCRAM-SHA-1": (AuthenticationFailed) Authentication failed.
    2024/10/09 10:02:48 Exit: connect to the node: ping: connection() error occurred during connection handshake: auth error: sasl conversation error: unable to authenticate using mechanism "SCRAM-SHA-1": (AuthenticationFailed) Authentication failed.
2024-10-09T10:02:48Z ERROR cannot start service: exited quickly with code 1
-----
```

As I said, it explodes because of pbm failing to start again.
Sequence of event was:
```
│ 12:02:58       │  start                              │
│ 12:02:46       │  config_changed                     │
│ 12:02:45       │  mongodb_logs_storage_attached      │
│ 12:02:44       │  mongodb_storage_attached           │
│ 12:02:43       │  secret_changed                     │
│ 12:02:38       │  mongod_pebble_ready                │
│ 12:02:35       │  leader_settings_changed            │
│ 12:02:35       │  database_peers_relation_created    │
│ 12:02:34       │  upgrade_version_a_relation_created │
│ 12:02:33       │  install                            │
```
So it started in `mongod_pebble_ready` and then `start` event tried to do the same and we end up in this situation.
I think somehow pebble is detecting a change and tries to restart the services.
I'll try to dig in to understand the difference as pebble has no logs for this check (looked up the code).

Reply:
After a discussion, and considering it's not the core of the PR here, we'll leave it as is and come back to it later in another PR.